Cancel statement doesn’t work due to “Session is locked by a concurrent client” (Regression for v2 client, v0.9.4)

<!--
    High Level Area Labels: 
     - client-api-v2 
     - client-v1
     - jdbc-v2

    Specific Area Labels:
      - packaging
      - sql-parser
      - data-type
      - docs
      - network
      - performance
      - jdbc-metadata
      - jdbc-insert
      - jdbc-read
      - client-insert
      - client-read
      - client-pojo-serde
      - format
     
-->
## Description
[This bug](https://github.com/ClickHouse/clickhouse-java/issues/1033) appears to have returned in the v2 version of the client. Short version: When you use a Session ID, `Statement.cancel()` never works. I expect you need the previous fix again, which removed the session ID from the `KILL QUERY` options.

### Steps to reproduce
1. Kick off a query that takes a long time and is killable using `Connection.createStatement()` and `Statement.executeUpdate(...)` or `Statement.execute()`, using a `clickhouse_setting_session_id`
2. Call `Statement.cancel()`
### Error Log or Exception StackTrace
```
java.sql.SQLException: com.clickhouse.client.api.ServerException: Code: 373. DB::Exception: Session e589cc67-36d5-4baa-bc94-86fe9fd120e3 is locked by a concurrent client. (SESSION_IS_LOCKED) (version 25.3.2.39 (official build)) 
	at com.clickhouse.jdbc.StatementImpl.cancel(StatementImpl.java:312) ~[clickhouse-jdbc-0.9.4-all.jar:clickhouse-jdbc 0.9.4 (revision: 39bd998)]
	at [My Code]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[?:?]
	at java.base/java.lang.reflect.Method.invoke(Method.java:580) ~[?:?]
	at [My Code]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[?:?]
	at java.base/java.lang.reflect.Method.invoke(Method.java:580) ~[?:?]
	at [My Code]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[?:?]
	at java.base/java.lang.reflect.Method.invoke(Method.java:580) ~[?:?]
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52) ~[jersey-server-2.35.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124) ~[jersey-server-2.35.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167) ~[jersey-server-2.35.jar:?]
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:219) ~[jersey-server-2.35.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79) ~[jersey-server-2.35.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:475) ~[jersey-server-2.35.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:397) ~[jersey-server-2.35.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:81) ~[jersey-server-2.35.jar:?]
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:255) ~[jersey-server-2.35.jar:?]
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248) ~[jersey-common-2.35.jar:?]
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244) ~[jersey-common-2.35.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292) ~[jersey-common-2.35.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274) ~[jersey-common-2.35.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244) ~[jersey-common-2.35.jar:?]
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265) ~[jersey-common-2.35.jar:?]
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:234) ~[jersey-server-2.35.jar:?]
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:684) ~[jersey-server-2.35.jar:?]
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:394) ~[jersey-container-servlet-core-2.35.jar:?]
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:346) ~[jersey-container-servlet-core-2.35.jar:?]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:366) ~[jersey-container-servlet-core-2.35.jar:?]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:319) ~[jersey-container-servlet-core-2.35.jar:?]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:205) ~[jersey-container-servlet-core-2.35.jar:?]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799) ~[jetty-servlet-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1656) ~[jetty-servlet-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.apache.shiro.web.servlet.AbstractShiroFilter.executeChain(AbstractShiroFilter.java:458) ~[shiro-web-1.13.0.jar:1.13.0]
	at org.apache.shiro.web.servlet.AbstractShiroFilter$1.call(AbstractShiroFilter.java:373) ~[shiro-web-1.13.0.jar:1.13.0]
	at org.apache.shiro.subject.support.SubjectCallable.doCall(SubjectCallable.java:90) ~[shiro-core-1.13.0.jar:1.13.0]
	at org.apache.shiro.subject.support.SubjectCallable.call(SubjectCallable.java:83) ~[shiro-core-1.13.0.jar:1.13.0]
	at org.apache.shiro.subject.support.DelegatingSubject.execute(DelegatingSubject.java:387) ~[shiro-core-1.13.0.jar:1.13.0]
	at org.apache.shiro.web.servlet.AbstractShiroFilter.doFilterInternal(AbstractShiroFilter.java:370) ~[shiro-web-1.13.0.jar:1.13.0]
	at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:154) ~[shiro-web-1.13.0.jar:1.13.0]
	at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193) ~[jetty-servlet-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1626) ~[jetty-servlet-9.4.58.v20250814.jar:9.4.58.v20250814]
	at com.liveaction.lanet.server.web.filters.MultiReadServletFilter.doFilter(MultiReadServletFilter.java:49) ~[main/:?]
	at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193) ~[jetty-servlet-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1626) ~[jetty-servlet-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:552) ~[jetty-servlet-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143) ~[jetty-server-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:600) ~[jetty-security-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127) ~[jetty-server-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235) ~[jetty-server-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624) ~[jetty-server-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233) ~[jetty-server-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440) ~[jetty-server-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188) ~[jetty-server-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:505) ~[jetty-servlet-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594) ~[jetty-server-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186) ~[jetty-server-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355) ~[jetty-server-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) ~[jetty-server-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146) ~[jetty-server-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127) ~[jetty-server-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.server.Server.handle(Server.java:516) ~[jetty-server-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487) ~[jetty-server-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732) ~[jetty-server-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479) ~[jetty-server-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277) ~[jetty-server-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311) ~[jetty-io-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105) ~[jetty-io-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.onFillable(SslConnection.java:555) ~[jetty-io-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:410) ~[jetty-io-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.io.ssl.SslConnection$2.succeeded(SslConnection.java:164) ~[jetty-io-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105) ~[jetty-io-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104) ~[jetty-io-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338) ~[jetty-util-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315) ~[jetty-util-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173) ~[jetty-util-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131) ~[jetty-util-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409) ~[jetty-util-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883) ~[jetty-util-9.4.58.v20250814.jar:9.4.58.v20250814]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034) ~[jetty-util-9.4.58.v20250814.jar:9.4.58.v20250814]
	at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
Caused by: com.clickhouse.client.api.ServerException: Code: 373. DB::Exception: Session e589cc67-36d5-4baa-bc94-86fe9fd120e3 is locked by a concurrent client. (SESSION_IS_LOCKED) (version 25.3.2.39 (official build)) 
	at com.clickhouse.client.api.internal.HttpAPIClientHelper.readError(HttpAPIClientHelper.java:403) ~[clickhouse-jdbc-0.9.4-all.jar:clickhouse-jdbc 0.9.4 (revision: 39bd998)]
	at com.clickhouse.client.api.internal.HttpAPIClientHelper.executeRequest(HttpAPIClientHelper.java:467) ~[clickhouse-jdbc-0.9.4-all.jar:clickhouse-jdbc 0.9.4 (revision: 39bd998)]
	at com.clickhouse.client.api.Client.lambda$query$5(Client.java:1599) ~[clickhouse-jdbc-0.9.4-all.jar:clickhouse-jdbc 0.9.4 (revision: 39bd998)]
	at com.clickhouse.client.api.Client.runAsyncOperation(Client.java:2014) ~[clickhouse-jdbc-0.9.4-all.jar:clickhouse-jdbc 0.9.4 (revision: 39bd998)]
	at com.clickhouse.client.api.Client.query(Client.java:1644) ~[clickhouse-jdbc-0.9.4-all.jar:clickhouse-jdbc 0.9.4 (revision: 39bd998)]
	at com.clickhouse.client.api.Client.query(Client.java:1542) ~[clickhouse-jdbc-0.9.4-all.jar:clickhouse-jdbc 0.9.4 (revision: 39bd998)]
	at com.clickhouse.jdbc.StatementImpl.cancel(StatementImpl.java:307) ~[clickhouse-jdbc-0.9.4-all.jar:clickhouse-jdbc 0.9.4 (revision: 39bd998)]
	... 98 more
``` 

### Expected Behaviour
"Killed query [UUID]" in log and query no longer running

### Code Example
Not sure how to give a concise example. Hopefully you already have a test for canceling queries.

### Configuration
Using `clickhouse_setting_session_id` (see linked bug ticket). May also need a session timeout, but don't think so.

#### Client Configuration
Using `clickhouse_setting_session_id` (see linked bug ticket). May also need a session timeout, but don't think so.

#### Environment
* [ ] Cloud (no)
* Client version: 0.9.4; v2 client
* Language version: Java 21.0.5+11
* OS: Windows 10

#### ClickHouse Server
* ClickHouse Server version: 25.3.2.39
* ClickHouse Server non-default settings, if any: N/A
* `CREATE TABLE` statements for tables involved: Anything that takes long enough to read that you can cancel it
* Sample data for all these tables, use [clickhouse-obfuscator](https://github.com/ClickHouse/ClickHouse/blob/master/programs/obfuscator/Obfuscator.cpp#L42-L80) if necessary

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,4 +1,4 @@
-## 0.10.9
+## 0.10.0
 
 ### Breaking Changes
 
@@ -38,6 +38,8 @@
 - **[jdbc-v2]** `ResultSet#getObject(int|String, Map<String, Class<?>>)` now accepts ClickHouse type names as map keys in addition to the JDBC `SQLType` names it has always accepted. Only unwrapped type names are used for the lookup — `Nullable(...)` and `LowCardinality(...)` wrappers are stripped and do not affect resolution, so a key like `"Int32"` matches both `Int32` and `Nullable(Int32)` columns; keys like `"Nullable(Int32)"` are not recognized. Lookup order is the `ClickHouseDataType` enum name (e.g. `"Int32"`, `"String"`, `"DateTime"`) then the JDBC `SQLType` name (e.g. `"INTEGER"`, `"VARCHAR"`, `"TIMESTAMP"`); a missing entry leaves the value uncoerced. The feature is supported for primitive ClickHouse types only — `Array`, `Tuple`, `Map`, `Nested`, and geometry types are not supported and continue to be returned in their native form regardless of the user-supplied map. Existing maps keyed only by JDBC `SQLType` names continue to work unchanged.
 
 ### Bug Fixes
+
+- **[jdbc-v2]** Fixed `Statement.cancel()` throwing `SESSION_IS_LOCKED` when the statement was running inside a ClickHouse session (e.g. via `clickhouse_setting_session_id`). The `KILL QUERY` request issued by `cancel()` now runs outside the session, so it no longer contends with the running query for the session lock. (https://github.com/ClickHouse/clickhouse-java/issues/2690)
 
 - **[client-v2]** Fixed inconsistent use of `executionTimeout` parameter in `Client` component. The timeout was previously set in milliseconds but mistakenly retrieved and used in seconds in some places. Now it correctly uses milliseconds consistently. (https://github.com/ClickHouse/clickhouse-java/issues/2358)
 

--- a/client-v2/src/main/java/com/clickhouse/client/api/command/CommandSettings.java
+++ b/client-v2/src/main/java/com/clickhouse/client/api/command/CommandSettings.java
@@ -34,6 +34,7 @@ public class CommandSettings extends QuerySettings {
         return this;
     }
 
+    @Override
     public CommandSettings clearSession() {
         super.clearSession();
         return this;

--- a/client-v2/src/main/java/com/clickhouse/client/api/command/CommandSettings.java
+++ b/client-v2/src/main/java/com/clickhouse/client/api/command/CommandSettings.java
@@ -33,4 +33,9 @@ public class CommandSettings extends QuerySettings {
         super.use(session);
         return this;
     }
+
+    public CommandSettings clearSession() {
+        super.clearSession();
+        return this;
+    }
 }

--- a/client-v2/src/main/java/com/clickhouse/client/api/insert/InsertSettings.java
+++ b/client-v2/src/main/java/com/clickhouse/client/api/insert/InsertSettings.java
@@ -4,7 +4,6 @@ import com.clickhouse.client.api.Client;
 import com.clickhouse.client.api.ClientConfigProperties;
 import com.clickhouse.client.api.Session;
 import com.clickhouse.client.api.internal.CommonSettings;
-import com.clickhouse.client.api.query.QuerySettings;
 import org.apache.hc.core5.http.HttpHeaders;
 
 import java.time.temporal.ChronoUnit;

--- a/client-v2/src/main/java/com/clickhouse/client/api/insert/InsertSettings.java
+++ b/client-v2/src/main/java/com/clickhouse/client/api/insert/InsertSettings.java
@@ -4,6 +4,7 @@ import com.clickhouse.client.api.Client;
 import com.clickhouse.client.api.ClientConfigProperties;
 import com.clickhouse.client.api.Session;
 import com.clickhouse.client.api.internal.CommonSettings;
+import com.clickhouse.client.api.query.QuerySettings;
 import org.apache.hc.core5.http.HttpHeaders;
 
 import java.time.temporal.ChronoUnit;
@@ -142,6 +143,11 @@ public class InsertSettings {
 
     public InsertSettings use(Session session) {
         settings.use(session);
+        return this;
+    }
+
+    public InsertSettings clearSession() {
+        settings.clearSession();
         return this;
     }
 

--- a/client-v2/src/main/java/com/clickhouse/client/api/internal/CommonSettings.java
+++ b/client-v2/src/main/java/com/clickhouse/client/api/internal/CommonSettings.java
@@ -143,7 +143,8 @@ public class CommonSettings {
         resetOption(ClientConfigProperties.serverSetting(ClickHouseHttpProto.QPARAM_SESSION_ID));
         resetOption(ClientConfigProperties.serverSetting(ClickHouseHttpProto.QPARAM_SESSION_CHECK));
         resetOption(ClientConfigProperties.serverSetting(ClickHouseHttpProto.QPARAM_SESSION_TIMEOUT));
-        resetOption(ClientConfigProperties.serverSetting(ClickHouseHttpProto.QPARAM_SESSION_TIMEZONE));
+        // Do not clean `session_timezone` setting because it is not related to session management and used to
+        // set timezone for consequent queries in some multi-user applications.
     }
 
     /**

--- a/client-v2/src/main/java/com/clickhouse/client/api/internal/CommonSettings.java
+++ b/client-v2/src/main/java/com/clickhouse/client/api/internal/CommonSettings.java
@@ -143,6 +143,7 @@ public class CommonSettings {
         resetOption(ClientConfigProperties.serverSetting(ClickHouseHttpProto.QPARAM_SESSION_ID));
         resetOption(ClientConfigProperties.serverSetting(ClickHouseHttpProto.QPARAM_SESSION_CHECK));
         resetOption(ClientConfigProperties.serverSetting(ClickHouseHttpProto.QPARAM_SESSION_TIMEOUT));
+        resetOption(ClientConfigProperties.serverSetting(ClickHouseHttpProto.QPARAM_SESSION_TIMEZONE));
     }
 
     /**

--- a/client-v2/src/main/java/com/clickhouse/client/api/internal/CommonSettings.java
+++ b/client-v2/src/main/java/com/clickhouse/client/api/internal/CommonSettings.java
@@ -139,6 +139,12 @@ public class CommonSettings {
         return this;
     }
 
+    public void clearSession() {
+        resetOption(ClientConfigProperties.serverSetting(ClickHouseHttpProto.QPARAM_SESSION_ID));
+        resetOption(ClientConfigProperties.serverSetting(ClickHouseHttpProto.QPARAM_SESSION_CHECK));
+        resetOption(ClientConfigProperties.serverSetting(ClickHouseHttpProto.QPARAM_SESSION_TIMEOUT));
+    }
+
     /**
      * Operation id. Used internally to register new operation.
      * Should not be called directly.

--- a/client-v2/src/main/java/com/clickhouse/client/api/query/QuerySettings.java
+++ b/client-v2/src/main/java/com/clickhouse/client/api/query/QuerySettings.java
@@ -139,6 +139,11 @@ public class QuerySettings {
         return this;
     }
 
+    public QuerySettings clearSession() {
+        settings.clearSession();
+        return this;
+    }
+
     /**
      * Read buffer is used for reading data from a server. Size is in bytes.
      * Minimal value is {@value MINIMAL_READ_BUFFER_SIZE} bytes.

--- a/client-v2/src/test/java/com/clickhouse/client/SettingsTests.java
+++ b/client-v2/src/test/java/com/clickhouse/client/SettingsTests.java
@@ -143,6 +143,30 @@ public class SettingsTests {
             Assert.assertEquals(settings.getSessionTimeout().intValue(), 45);
             Assert.assertEquals(settings.getSessionTimezone(), "Europe/Berlin");
         }
+
+        {
+            final QuerySettings settings = new QuerySettings();
+            settings.setSessionId("session-clear-1");
+            settings.setSessionCheck(true);
+            settings.setSessionTimeout(60);
+            settings.setSessionTimezone("America/New_York");
+            Assert.assertNotNull(settings.getSessionId());
+            Assert.assertNotNull(settings.getSessionCheck());
+            Assert.assertNotNull(settings.getSessionTimeout());
+            Assert.assertNotNull(settings.getSessionTimezone());
+
+            settings.clearSession();
+
+            Assert.assertNull(settings.getSessionId(), "clearSession() must remove session_id");
+            Assert.assertNull(settings.getSessionCheck(), "clearSession() must remove session_check");
+            Assert.assertNull(settings.getSessionTimeout(), "clearSession() must remove session_timeout");
+            Assert.assertNull(settings.getSessionTimezone(), "clearSession() must remove session_timezone");
+
+            // Non-session settings are unaffected.
+            settings.setDatabase("db1");
+            settings.clearSession();
+            Assert.assertEquals(settings.getDatabase(), "db1");
+        }
     }
 
     @Test
@@ -231,6 +255,30 @@ public class SettingsTests {
             Assert.assertFalse(settings.getSessionCheck());
             Assert.assertEquals(settings.getSessionTimeout().intValue(), 50);
             Assert.assertEquals(settings.getSessionTimezone(), "Europe/Paris");
+        }
+
+        {
+            final InsertSettings settings = new InsertSettings();
+            settings.setSessionId("session-clear-2");
+            settings.setSessionCheck(true);
+            settings.setSessionTimeout(90);
+            settings.setSessionTimezone("Asia/Tokyo");
+            Assert.assertNotNull(settings.getSessionId());
+            Assert.assertNotNull(settings.getSessionCheck());
+            Assert.assertNotNull(settings.getSessionTimeout());
+            Assert.assertNotNull(settings.getSessionTimezone());
+
+            settings.clearSession();
+
+            Assert.assertNull(settings.getSessionId(), "clearSession() must remove session_id");
+            Assert.assertNull(settings.getSessionCheck(), "clearSession() must remove session_check");
+            Assert.assertNull(settings.getSessionTimeout(), "clearSession() must remove session_timeout");
+            Assert.assertNull(settings.getSessionTimezone(), "clearSession() must remove session_timezone");
+
+            // Non-session settings are unaffected.
+            settings.setDatabase("db2");
+            settings.clearSession();
+            Assert.assertEquals(settings.getDatabase(), "db2");
         }
     }
 }

--- a/client-v2/src/test/java/com/clickhouse/client/SettingsTests.java
+++ b/client-v2/src/test/java/com/clickhouse/client/SettingsTests.java
@@ -160,7 +160,9 @@ public class SettingsTests {
             Assert.assertNull(settings.getSessionId(), "clearSession() must remove session_id");
             Assert.assertNull(settings.getSessionCheck(), "clearSession() must remove session_check");
             Assert.assertNull(settings.getSessionTimeout(), "clearSession() must remove session_timeout");
-            Assert.assertNull(settings.getSessionTimezone(), "clearSession() must remove session_timezone");
+            // session_timezone is not session-management state; it is preserved across clearSession().
+            Assert.assertEquals(settings.getSessionTimezone(), "America/New_York",
+                    "clearSession() must not remove session_timezone");
 
             // Non-session settings are unaffected.
             settings.setDatabase("db1");
@@ -273,7 +275,9 @@ public class SettingsTests {
             Assert.assertNull(settings.getSessionId(), "clearSession() must remove session_id");
             Assert.assertNull(settings.getSessionCheck(), "clearSession() must remove session_check");
             Assert.assertNull(settings.getSessionTimeout(), "clearSession() must remove session_timeout");
-            Assert.assertNull(settings.getSessionTimezone(), "clearSession() must remove session_timezone");
+            // session_timezone is not session-management state; it is preserved across clearSession().
+            Assert.assertEquals(settings.getSessionTimezone(), "Asia/Tokyo",
+                    "clearSession() must not remove session_timezone");
 
             // Non-session settings are unaffected.
             settings.setDatabase("db2");

--- a/jdbc-v2/src/main/java/com/clickhouse/jdbc/StatementImpl.java
+++ b/jdbc-v2/src/main/java/com/clickhouse/jdbc/StatementImpl.java
@@ -334,7 +334,7 @@ public class StatementImpl implements Statement, JdbcV2Wrapper {
             return;
         }
 
-        // KILL QUERY must not run inside the same session as the query being cancelled otherwise it wil
+        // KILL QUERY must not run inside the same session as the query being canceled otherwise it will
         // cause "Session is locked by a concurrent client" (SESSION_IS_LOCKED) error.
         QuerySettings cancelSettings = QuerySettings.merge(getLocalSettings(), new QuerySettings()).clearSession();
         try (QueryResponse response = connection.getClient().query(String.format("KILL QUERY%sWHERE query_id = '%s'",

--- a/jdbc-v2/src/main/java/com/clickhouse/jdbc/StatementImpl.java
+++ b/jdbc-v2/src/main/java/com/clickhouse/jdbc/StatementImpl.java
@@ -2,6 +2,7 @@ package com.clickhouse.jdbc;
 
 import com.clickhouse.client.api.ClientConfigProperties;
 import com.clickhouse.client.api.data_formats.ClickHouseBinaryFormatReader;
+import com.clickhouse.client.api.http.ClickHouseHttpProto;
 import com.clickhouse.client.api.internal.ServerSettings;
 import com.clickhouse.client.api.query.QueryResponse;
 import com.clickhouse.client.api.query.QuerySettings;
@@ -333,9 +334,12 @@ public class StatementImpl implements Statement, JdbcV2Wrapper {
             return;
         }
 
+        // KILL QUERY must not run inside the same session as the query being cancelled otherwise it wil
+        // cause "Session is locked by a concurrent client" (SESSION_IS_LOCKED) error.
+        QuerySettings cancelSettings = QuerySettings.merge(getLocalSettings(), new QuerySettings()).clearSession();
         try (QueryResponse response = connection.getClient().query(String.format("KILL QUERY%sWHERE query_id = '%s'",
                 connection.onCluster ? " ON CLUSTER " + SQLUtils.enquoteIdentifier(connection.cluster, true) + ' ' : ' ',
-                lastQueryId), connection.getDefaultQuerySettings()).get()){
+                lastQueryId), cancelSettings).get()){
             LOG.debug("Query {} was killed by {}", lastQueryId, response.getQueryId());
         } catch (Exception e) {
             throw new SQLException(e);

--- a/jdbc-v2/src/main/java/com/clickhouse/jdbc/StatementImpl.java
+++ b/jdbc-v2/src/main/java/com/clickhouse/jdbc/StatementImpl.java
@@ -2,7 +2,6 @@ package com.clickhouse.jdbc;
 
 import com.clickhouse.client.api.ClientConfigProperties;
 import com.clickhouse.client.api.data_formats.ClickHouseBinaryFormatReader;
-import com.clickhouse.client.api.http.ClickHouseHttpProto;
 import com.clickhouse.client.api.internal.ServerSettings;
 import com.clickhouse.client.api.query.QueryResponse;
 import com.clickhouse.client.api.query.QuerySettings;

--- a/jdbc-v2/src/test/java/com/clickhouse/jdbc/StatementTest.java
+++ b/jdbc-v2/src/test/java/com/clickhouse/jdbc/StatementTest.java
@@ -9,7 +9,6 @@ import org.apache.commons.lang3.RandomStringUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.testng.Assert;
-import org.testng.SkipException;
 import org.testng.annotations.DataProvider;
 import org.testng.annotations.Test;
 
@@ -29,6 +28,8 @@ import java.util.Map;
 import java.util.Properties;
 import java.util.UUID;
 import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
 
 import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.assertFalse;
@@ -706,6 +707,126 @@ public class StatementTest extends JdbcIntegrationTest {
 
                     latch.await();
                     stmt.cancel();
+                }
+            }
+        }
+    }
+
+    /**
+     * Waits until the given query id appears in {@code system.processes}, so we know the operation has actually
+     * started on the server before attempting to cancel it. Uses a dedicated connection (no session) to observe.
+     */
+    private boolean waitForQueryToStart(String queryId, int timeoutSeconds) throws Exception {
+        if (queryId == null) {
+            return false;
+        }
+        long deadline = System.currentTimeMillis() + TimeUnit.SECONDS.toMillis(timeoutSeconds);
+        while (System.currentTimeMillis() < deadline) {
+            try (Connection conn = getJdbcConnection();
+                 Statement stmt = conn.createStatement();
+                 ResultSet rs = stmt.executeQuery("SELECT count() FROM system.processes WHERE query_id = '" + queryId + "'")) {
+
+                if (rs.next() && rs.getLong(1) > 0) {
+                    return true;
+                }
+            }
+            Thread.sleep(200);
+        }
+        return false;
+    }
+
+    private String waitForQueryId(StatementImpl stmt, int timeoutSeconds) throws Exception {
+        long deadline = System.currentTimeMillis() + TimeUnit.SECONDS.toMillis(timeoutSeconds);
+        while (System.currentTimeMillis() < deadline) {
+            String queryId = stmt.getLastQueryId();
+            if (queryId != null && !queryId.isEmpty()) {
+                return queryId;
+            }
+            Thread.sleep(50);
+        }
+        return null;
+    }
+
+    @Test(groups = {"integration"})
+    public void testCancelQueryWithSession() throws Exception {
+        // Regression test for #2690: cancelling a query that runs inside a session must not fail with
+        // "Session is locked by a concurrent client" (SESSION_IS_LOCKED). The KILL QUERY request issued by
+        // cancel() must not carry the session id of the query being cancelled.
+        String sessionId = "test-session-" + UUID.randomUUID();
+        try (Connection conn = getJdbcConnection()) {
+            try (StatementImpl stmt = (StatementImpl) conn.createStatement()) {
+                stmt.getLocalSettings().setSessionId(sessionId);
+                stmt.setQueryTimeout(30); // safety net so a failed cancel cannot hang the test
+
+                final AtomicReference<Throwable> threadError = new AtomicReference<>();
+                final CountDownLatch started = new CountDownLatch(1);
+                Thread worker = new Thread(() -> {
+                    started.countDown();
+                    // Long-running query that only completes when killed.
+                    try (ResultSet rs = stmt.executeQuery("SELECT count() FROM system.numbers_mt")) {
+                        rs.next();
+                    } catch (Throwable t) {
+                        System.out.println("Error: " + t.getMessage());
+                        threadError.set(t);
+                    }
+                });
+                worker.start();
+                started.await();
+
+                String queryId = waitForQueryId(stmt, 15);
+                assertNotNull(queryId, "Query id was not assigned in time");
+                assertTrue(waitForQueryToStart(queryId, 15), "Query did not start on the server in time");
+
+                // Cancel from the main thread - must not throw SESSION_IS_LOCKED.
+                stmt.cancel();
+
+                worker.join(TimeUnit.SECONDS.toMillis(20));
+                assertFalse(worker.isAlive(), "Query was not cancelled and is still running");
+            }
+        }
+    }
+
+    @Test(groups = {"integration"})
+    public void testCancelInsertWithSession() throws Exception {
+        // Regression test for #2690 covering a long-running INSERT executed inside a session.
+        String tableName = getDatabase() + ".cancel_insert_with_session";
+        String sessionId = "test-session-" + UUID.randomUUID();
+        try (Connection conn = getJdbcConnection(Map.of(ASYNC_INSERT_SETTING_KEY, ServerSettings.OFF))) {
+            try (Statement setup = conn.createStatement()) {
+                setup.execute("DROP TABLE IF EXISTS " + tableName);
+                setup.execute("CREATE TABLE " + tableName + " (num UInt64) ENGINE = MergeTree ORDER BY ()");
+            }
+
+            try (StatementImpl stmt = (StatementImpl) conn.createStatement()) {
+                stmt.getLocalSettings().setSessionId(sessionId);
+                stmt.setQueryTimeout(30); // safety net so a failed cancel cannot hang the test
+
+                final AtomicReference<Throwable> threadError = new AtomicReference<>();
+                final CountDownLatch started = new CountDownLatch(1);
+                Thread worker = new Thread(() -> {
+                    started.countDown();
+                    // Long-running insert that only completes when killed.
+                    try {
+                        stmt.executeUpdate("INSERT INTO " + tableName + " SELECT number FROM system.numbers_mt");
+                    } catch (Throwable t) {
+                        threadError.set(t);
+                    }
+                });
+                worker.start();
+                started.await();
+
+                String queryId = waitForQueryId(stmt, 15);
+                assertNotNull(queryId, "Query id was not assigned in time");
+                assertTrue(waitForQueryToStart(queryId, 15), "Insert did not start on the server in time");
+
+                // Cancel from the main thread - must not throw SESSION_IS_LOCKED.
+                stmt.cancel();
+
+                worker.join(TimeUnit.SECONDS.toMillis(20));
+                assertFalse(worker.isAlive(), "Insert was not cancelled and is still running");
+            } finally {
+                try (Statement cleanup = conn.createStatement()) {
+                    cleanup.execute("DROP TABLE IF EXISTS " + tableName);
                 }
             }
         }

--- a/jdbc-v2/src/test/java/com/clickhouse/jdbc/StatementTest.java
+++ b/jdbc-v2/src/test/java/com/clickhouse/jdbc/StatementTest.java
@@ -9,6 +9,7 @@ import org.apache.commons.lang3.RandomStringUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.testng.Assert;
+import org.testng.SkipException;
 import org.testng.annotations.DataProvider;
 import org.testng.annotations.Test;
 
@@ -749,6 +750,10 @@ public class StatementTest extends JdbcIntegrationTest {
 
     @Test(groups = {"integration"})
     public void testCancelQueryWithSession() throws Exception {
+        if (isCloud()) {
+            throw new SkipException("Cloud + HTTP doesn't work well. Enough to test locally");
+        }
+
         // Regression test for #2690: cancelling a query that runs inside a session must not fail with
         // "Session is locked by a concurrent client" (SESSION_IS_LOCKED). The KILL QUERY request issued by
         // cancel() must not carry the session id of the query being cancelled.
@@ -788,6 +793,9 @@ public class StatementTest extends JdbcIntegrationTest {
 
     @Test(groups = {"integration"})
     public void testCancelInsertWithSession() throws Exception {
+        if (isCloud()) {
+            throw new SkipException("Cloud + HTTP doesn't work well. Enough to test locally");
+        }
         // Regression test for #2690 covering a long-running INSERT executed inside a session.
         String tableName = getDatabase() + ".cancel_insert_with_session";
         String sessionId = "test-session-" + UUID.randomUUID();

